Support GPUAutoLayoutMode

mwyrzykowski opened the pull request:
#### 60342c994bcb6d900091f653662773eb1db91fe7
<pre>
Support GPUAutoLayoutMode
<a href="https://bugs.webkit.org/show_bug.cgi?id=248706">https://bugs.webkit.org/show_bug.cgi?id=248706</a>
&lt;rdar://101404965&gt;

Reviewed by Dean Jackson.

Allow layout: &quot;auto&quot; to be specified in place of creating a pipeline layout.

This is needed to support running some WebGPU existing samples, e.g., the ones
located at <a href="https://austin-eng.com/webgpu-samples/">https://austin-eng.com/webgpu-samples/</a>

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/WebGPU/GPUAutoLayoutMode.h: Copied from Source/WebCore/Modules/WebGPU/GPUShaderModuleCompilationHint.idl.
* Source/WebCore/Modules/WebGPU/GPUAutoLayoutMode.idl: Copied from Source/WebCore/Modules/WebGPU/GPUPipelineDescriptorBase.idl.
Add files.
* Source/WebCore/Modules/WebGPU/GPUComputePipelineDescriptor.h:
(WebCore::GPUComputePipelineDescriptor::convertToBacking const):
* Source/WebCore/Modules/WebGPU/GPUDevice.cpp:
(WebCore::GPUDevice::GPUDevice):
(WebCore::m_queue):
(WebCore::m_autoPipelineLayout):
(WebCore::GPUDevice::createShaderModule):
(WebCore::GPUDevice::createComputePipeline):
(WebCore::GPUDevice::createRenderPipeline):
(WebCore::GPUDevice::createComputePipelineAsync):
(WebCore::GPUDevice::createRenderPipelineAsync):
Support auto pipeline layouts by passing an empty pipeline layout.

* Source/WebCore/Modules/WebGPU/GPUDevice.h:
(WebCore::GPUDevice::GPUDevice): Deleted.
(WebCore::GPUDevice::m_queue): Deleted.
Moved to cpp file.

* Source/WebCore/Modules/WebGPU/GPUPipelineDescriptorBase.h:
(WebCore::convertPipelineLayoutToBacking):
(WebCore::GPUPipelineDescriptorBase::convertToBacking const):
* Source/WebCore/Modules/WebGPU/GPUPipelineDescriptorBase.idl:
* Source/WebCore/Modules/WebGPU/GPURenderPipelineDescriptor.h:
(WebCore::GPURenderPipelineDescriptor::convertToBacking const):
* Source/WebCore/Modules/WebGPU/GPUShaderModuleCompilationHint.h:
(WebCore::GPUShaderModuleCompilationHint::convertToBacking const):
* Source/WebCore/Modules/WebGPU/GPUShaderModuleCompilationHint.idl:
* Source/WebCore/Modules/WebGPU/GPUShaderModuleDescriptor.h:
(WebCore::GPUShaderModuleDescriptor::convertToBacking const):
Pass the auto layout to convertToBacking(), so we can fall back to that
one if needed.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Websites/webkit.org/demos/webgpu/hello-cube-auto-layout.html: Added.
* Websites/webkit.org/demos/webgpu/scripts/hello-cube-auto-layout.js: Added.
(async helloCube.frameUpdate):
(async helloCube):
Example using layout: &quot;auto&quot; in the call to createRenderPipeline and the shader
hint, which are both supported in the specification.

Canonical link: <a href="https://commits.webkit.org/257537@main">https://commits.webkit.org/257537@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/402bc8ebc69a3ba2a2f7549ed87a633a6323088b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99196 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8404 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32330 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108589 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168836 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/103195 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8962 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85729 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91698 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106523 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104941 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/6766 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90327 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33771 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/88586 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21674 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/76638 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2289 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23190 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2179 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/45592 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5179 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/7233 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42663 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3656 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

grorg commented on Source/WebCore/Modules/WebGPU/GPUDevice.cpp:
```diff
@@ -68,6 +68,14 @@ namespace WebCore {
 WTF_MAKE_ISO_ALLOCATED_IMPL(GPUDevice);
 
+GPUDevice::GPUDevice(ScriptExecutionContext* scriptExecutionContext, Ref<PAL::WebGPU::Device>&& backing)
+: ActiveDOMObject { scriptExecutionContext }
```
I think these are normally indented.

mwyrzykowski replied:
I thought that was the case too, but when I indent style checker says:

```ERROR: Source/WebCore/Modules/WebGPU/GPUDevice.cpp:73:  Code inside a namespace should not be indented```

should I ignore the style checker warning? Diff is this:

```
GPUDevice::GPUDevice(ScriptExecutionContext* scriptExecutionContext, Ref<PAL::WebGPU::Device>&& backing)
-: ActiveDOMObject { scriptExecutionContext }
-, m_backing(WTFMove(backing))
-, m_queue(GPUQueue::create(Ref { m_backing->queue() }))
-, m_autoPipelineLayout(createPipelineLayout({ { "autoLayout"_s, }, { } }))
+    : ActiveDOMObject { scriptExecutionContext }
+    , m_backing(WTFMove(backing))
+    , m_queue(GPUQueue::create(Ref { m_backing->queue() }))
+    , m_autoPipelineLayout(createPipelineLayout({ { "autoLayout"_s, }, { } }))
 {
 }
```

mwyrzykowski replied:
Clarified offline that I should ignore style checker here

djg replied:
I'm pretty sure that the initializer list should be indented.

mwyrzykowski replied:
Hopefully I can still merge it :) https://ews-build.webkit.org/#/builders/6/builds/99196

mwyrzykowski committed:
Support GPUAutoLayoutMode
https://bugs.webkit.org/show_bug.cgi?id=248706
<rdar://101404965>

Reviewed by Dean Jackson.

Allow layout: "auto" to be specified in place of creating a pipeline layout.

This is needed to support running some WebGPU existing samples, e.g., the ones
located at https://austin-eng.com/webgpu-samples/

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/WebGPU/GPUAutoLayoutMode.h: Copied from Source/WebCore/Modules/WebGPU/GPUShaderModuleCompilationHint.idl.
* Source/WebCore/Modules/WebGPU/GPUAutoLayoutMode.idl: Copied from Source/WebCore/Modules/WebGPU/GPUPipelineDescriptorBase.idl.
Add files.
* Source/WebCore/Modules/WebGPU/GPUComputePipelineDescriptor.h:
(WebCore::GPUComputePipelineDescriptor::convertToBacking const):
* Source/WebCore/Modules/WebGPU/GPUDevice.cpp:
(WebCore::GPUDevice::GPUDevice):
(WebCore::m_queue):
(WebCore::m_autoPipelineLayout):
(WebCore::GPUDevice::createShaderModule):
(WebCore::GPUDevice::createComputePipeline):
(WebCore::GPUDevice::createRenderPipeline):
(WebCore::GPUDevice::createComputePipelineAsync):
(WebCore::GPUDevice::createRenderPipelineAsync):
Support auto pipeline layouts by passing an empty pipeline layout.

* Source/WebCore/Modules/WebGPU/GPUDevice.h:
(WebCore::GPUDevice::GPUDevice): Deleted.
(WebCore::GPUDevice::m_queue): Deleted.
Moved to cpp file.

* Source/WebCore/Modules/WebGPU/GPUPipelineDescriptorBase.h:
(WebCore::convertPipelineLayoutToBacking):
(WebCore::GPUPipelineDescriptorBase::convertToBacking const):
* Source/WebCore/Modules/WebGPU/GPUPipelineDescriptorBase.idl:
* Source/WebCore/Modules/WebGPU/GPURenderPipelineDescriptor.h:
(WebCore::GPURenderPipelineDescriptor::convertToBacking const):
* Source/WebCore/Modules/WebGPU/GPUShaderModuleCompilationHint.h:
(WebCore::GPUShaderModuleCompilationHint::convertToBacking const):
* Source/WebCore/Modules/WebGPU/GPUShaderModuleCompilationHint.idl:
* Source/WebCore/Modules/WebGPU/GPUShaderModuleDescriptor.h:
(WebCore::GPUShaderModuleDescriptor::convertToBacking const):
Pass the auto layout to convertToBacking(), so we can fall back to that
one if needed.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Websites/webkit.org/demos/webgpu/hello-cube-auto-layout.html: Added.
* Websites/webkit.org/demos/webgpu/scripts/hello-cube-auto-layout.js: Added.
(async helloCube.frameUpdate):
(async helloCube):
Example using layout: "auto" in the call to createRenderPipeline and the shader
hint, which are both supported in the specification.

Canonical link: https://commits.webkit.org/257537@main
